src/client/Client.h

```diff
@@ -750,13 +751,22 @@ class Client : public Dispatcher, public md_config_obs_t {
 
   xlist<Inode*> &get_dirty_list() { return dirty_list; }
 
+  void set_mdsmap(const bufferlist &bl) {
+    std::shared_ptr<MDSMap> _mdsmap(new MDSMap);
+    _mdsmap->decode(bl);
+    atomic_store_explicit(&mdsmap, std::move(_mdsmap), std::memory_order_relaxed);
+  }
+  std::shared_ptr<MDSMap> get_mdsmap() const {
+    auto _mdsmap = atomic_load_explicit(&mdsmap, std::memory_order_relaxed);
+    return _mdsmap;
+  }
```

Comment:
I'm thinking about this more @lxbsz and I'm not sure we can actually do this, unfortunately. If we have multiple Client threads loading this MDSMap, we could have one thread writing out a request based on the contents of the old map **after** another thread has loaded the newmap and written a request based on the newmap.

There are ways to resolve this however. I think the fastest way to address this is via a `std::shared_mutex`. The `handle_mds_map` would just acquire a unique_lock to update the mdsmap. The `std::shared_lock` would be made part of your `RWRef_t`(as a wrapper class) that also keeps a reader lock on the mdsmap. IOW: I think the client entry points needs to setup state/locks that it passes down the stack. This includes the lock on the mount state (your `RWRef_t`), the mdsmap state, and any other  potential new state.

Note, in the end something other than shared_lock would be preferable because we would like to have the mdsmap handler acquire the unique_lock with priority (new readers would defer). That could be done by keeping track of the epoch readers are using and the epoch that the writer (`handle_mds_map`) wants to write. If a new reader sees the epochs do not match, then it waits. For now however, I think a `shared_lock` is suitable because it is starvation free in practice [1].

[1] http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2406.html#shared_mutex_imp

Reply:
> I'm thinking about this more @lxbsz and I'm not sure we can actually do this, unfortunately. If we have multiple Client threads loading this MDSMap, we could have one thread writing out a request based on the contents of the old map **after** another thread has loaded the newmap and written a request based on the newmap.
> 

Right, it could happen. And tt could be a problem that when making the request if it is using the old mdsmap, while actually the new map is just come, it may try wait for the new map if the state is not active, which will miss the one it just received.

> There are ways to resolve this however. I think the fastest way to address this is via a `std::shared_mutex`. The `handle_mds_map` would just acquire a unique_lock to update the mdsmap. The `std::shared_lock` would be made part of your `RWRef_t`(as a wrapper class) that also keeps a reader lock on the mdsmap. IOW: I think the client entry points needs to setup state/locks that it passes down the stack. This includes the lock on the mount state (your `RWRef_t`), the mdsmap state, and any other potential new state.
> 

Yeah, this approach sounds good and simple.

> Note, in the end something other than shared_lock would be preferable because we would like to have the mdsmap handler acquire the unique_lock with priority (new readers would defer). That could be done by keeping track of the epoch readers are using and the epoch that the writer (`handle_mds_map`) wants to write. If a new reader sees the epochs do not match, then it waits. For now however, I think a `shared_lock` is suitable because it is starvation free in practice [1].
> 
> [1] http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2406.html#shared_mutex_imp

Reply:
@batrick I am not confident in the code locally and it is a little risky. So I need more time on it.
I have raised one new tricker https://tracker.ceph.com/issues/47920 to follow it.

For this PR let's just make it clean and for command_lock. I will work on the `mdsmap` commit in one separate PR later.